### PR TITLE
issue-2553 GetTipSetAndState should be split into two methods v0

### PR DESCRIPTION
**Pull request by jscode017**
My first try for this issue.
I think that there are some question remain:
1. How to name abbreviate tipsetstateroot?
I see that most of tipset in code is named ts, so I name tipsetstateroot tssr, but it seems weird. So maybe something like tsStateRoot?
2.At [here](https://github.com/jscode017/go-filecoin/blob/34c19cb529fb27c26aad974f3897ee4462f77f80/chain/tip_index.go#L79) What I did in chain/tip_index.go is that the Get() function remains and write to new methods to access Get() function then access TipSet and TipSetStateRoot respectively, I'm not sure is this appropriate?

### CODE REVIEW

File: chain/default_store.go [hunk 1]
```diff
@@ -384,11 +390,11 @@ func (store *DefaultStore) BlockHeight() (uint64, error) {
 
 // ActorFromLatestState gets the latest state and retrieves an actor from it.
 func (store *DefaultStore) ActorFromLatestState(ctx context.Context, addr address.Address) (*actor.Actor, error) {
-	tsas, err := store.GetTipSetAndState(store.GetHead())
+	tssr, err := store.GetTipSetStateRoot(store.GetHead())
```
Comment (jscode017) [inline chain/default_store.go]: maybe should change tssr to something like tsStateRoot?

File: plumbing/msg/queryer.go [hunk 2]
```diff
@@ -45,15 +45,19 @@ func (q *Queryer) Query(ctx context.Context, optFrom, to address.Address, method
...
 	if err != nil {
 		return nil, errors.Wrap(err, "couldnt get latest state root")
 	}
-	st, err := state.LoadStateTree(ctx, q.cst, tsas.TipSetStateRoot, builtin.Actors)
+	st, err := state.LoadStateTree(ctx, q.cst, tssr, builtin.Actors)
 	if err != nil {
 		return nil, errors.Wrap(err, "could load tree for latest state root")
 	}
-	h, err := tsas.TipSet.Height()
+	ts, err := q.chainReader.GetTipSet(headTs)
```
Comment (rosalinekarr) [inline plumbing/msg/queryer.go]: We can also use `BlockHeight()` here to simplify a bit.

File: chain/default_syncer.go [hunk 3]
```diff
@@ -159,11 +159,11 @@ func (syncer *DefaultSyncer) tipSetState(ctx context.Context, tsKey types.Sorted
 	if !syncer.chainStore.HasTipSetAndState(ctx, tsKey.String()) {
 		return nil, errors.Wrap(ErrUnexpectedStoreState, "parent tipset must be in the store")
 	}
-	tsas, err := syncer.chainStore.GetTipSetAndState(tsKey)
+	tssr, err := syncer.chainStore.GetTipSetStateRoot(tsKey)
 	if err != nil {
 		return nil, err
 	}
-	st, err := state.LoadStateTree(ctx, syncer.stateStore, tsas.TipSetStateRoot, builtin.Actors)
+	st, err := state.LoadStateTree(ctx, syncer.stateStore, tssr, builtin.Actors)
```
Comment (rosalinekarr) [inline chain/default_syncer.go]: Just a heads up for anyone else reviewing this, this logic and all the repetitions of it should be greatly simplified by subsequent work relating to issue #2552.

File: chain/default_store.go (hunk 1 above)
Comment (rosalinekarr) [reply chain/default_store.go]: That sounds reasonable, but I don't have any strong opinions either way. @anorth, do you?

File: chain/default_syncer.go [hunk 4]
```diff
@@ -260,8 +260,8 @@ func (syncer *DefaultSyncer) syncOne(ctx context.Context, parent, next types.Tip
 			return err
 		}
 		newChain = append(newChain, next)
-		if IsReorg(headTipSetAndState.TipSet, newChain) {
-			logSyncer.Infof("reorg occurring while switching from %s to %s", headTipSetAndState.TipSet.String(), next.String())
+		if IsReorg(*headTipSet, newChain) {
+			logSyncer.Infof("reorg occurring while switching from %s to %s", (*headTipSet).String(), next.String())
```
[the remark this answers is not in the document]
Comment (jscode017) [reply chain/default_syncer.go]: Yeah, but passing it into the function will get error?
cannot use headTipSet (type *types.TipSet) as type types.TipSet in argument to IsReorg

[the remark this answers is not in the document]
Comment (rosalinekarr) [reply chain/default_syncer.go]: yeah, line 263 is fine. i was suggesting changing line 264 to the following:

```
logSyncer.Infof("reorg occurring while switching from %s to %s", headTipSet.String(), next.String())
```

File: plumbing/msg/previewer.go [hunk 5]
```diff
@@ -42,15 +42,19 @@ func (p *Previewer) Preview(ctx context.Context, optFrom, to address.Address, me
...
 	if err != nil {
 		return types.NewGasUnits(0), errors.Wrap(err, "couldnt get latest state root")
 	}
-	st, err := state.LoadStateTree(ctx, p.cst, tsas.TipSetStateRoot, builtin.Actors)
+	st, err := state.LoadStateTree(ctx, p.cst, tssr, builtin.Actors)
 	if err != nil {
 		return types.NewGasUnits(0), errors.Wrap(err, "could load tree for latest state root")
 	}
-	h, err := tsas.TipSet.Height()
+	ts, err := p.chainReader.GetTipSet(headTs)
```
[the remark this answers is not in the document]
Comment (jscode017) [reply plumbing/msg/previewer.go]: This is somehow not in the ReadStore interface, so I add it in.

File: chain/store.go [hunk 6]
```diff
@@ -48,6 +48,9 @@ type ReadStore interface {
 	ActorFromLatestState(ctx context.Context, address address.Address) (*actor.Actor, error)
 
 	GenesisCid() cid.Cid
+
+	//returns the chain height of the head tipset
```
Comment (jscode017) [inline chain/store.go]: Add the BlockHeight() function in the store interface

File: chain/store.go [hunk 7]
```diff
@@ -48,6 +48,9 @@ type ReadStore interface {
...
+	BlockHeight() (uint64, error)
```
Comment (jscode017) [inline chain/store.go]: Add the BlockHeight() function in the store interface

File: chain/default_syncer.go (hunk 4 above)
[the remark this answers is not in the document]
Comment (jscode017) [reply chain/default_syncer.go]: OK, I removed the deference.

**Comment by rosalinekarr** [comment]
Looks good to me! You'll need rebase on the current master branch and get one more approval from @anort then I'll merge the changes.

**Comment by jscode017** [comment]
I rebase to master branch at this current time. If need any further rebase just that me know.

File: chain/default_store.go (hunk 1 above)
Comment (anorth) [reply chain/default_store.go]: Yes, `stateRoot` or `stateCid` would be better.

File: plumbing/msg/sender.go [hunk 8]
```diff
@@ -28,7 +28,8 @@ const Topic = "/fil/msgs"
 // Abstracts over a store of blockchain state.
 type chainState interface {
 	GetHead() types.SortedCidSet
-	GetTipSetAndState(tsKey types.SortedCidSet) (*chain.TipSetAndState, error)
+	GetTipSet(tsKey types.SortedCidSet) (*types.TipSet, error)
```
Comment (anorth) [inline plumbing/msg/sender.go]: The sender doesn't use `GetTipSet`, so please remove it from this interface.

**Comment by jscode017** [comment]
Can pass go run ./build best in my mac, however, somehow run into error here.

**Comment by rosalinekarr** [comment]
@jscode017 Tests seem to pass on my machine as well. It's possible CI may have just had an intermittent failure. Try re-running them and see if they pass on a second attempt.

If they do, we can go ahead and merge this then open up a new issue for the intermittent test failure.

**Comment by jscode017** [comment]
@rosalinekarr I think that I need write permission to rerun the CI test?

**Comment by rosalinekarr** [comment]
@jscode017 Thanks for your great contribution to the project!

**Pull request merged**